[AKV] Add new preview version for NSP feature

osmuller opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/20296335/9e51cff7-28f7-4ae7-99a3-0f4a116c759f)

<details>
<summary>Click here to see the details of Step 1, Breaking Changes review</summary>

If you are in purview of Step 1 of the diagram, follow the [Breaking Changes review process].  
**IMPORTANT!** This applies even if you believe your PR was mislabeled, for any reason, including tool failure.
</details>

<details> 
<summary>Click here to see the details of Step 2, ARM review</summary>

See https://aka.ms/azsdk/pr-arm-review.
</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+label%3Aresource-manager+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+label%3Aresource-manager+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[Breaking Changes review process]: https://aka.ms/brch

osmuller committed:
Start adding NSP changes to preview version

osmuller committed:
add example

osmuller committed:
Change to the response to be list instead of get

osmuller committed:
Add missing example file

osmuller committed:
Add missing enabledLogCategory

osmuller committed:
Fix typo

osmuller committed:
Test object type as array

osmuller committed:
Test nsp access rule change

osmuller committed:
Fixes for Lint

osmuller committed:
Add pageable

osmuller committed:
Make list not have a next link

osmuller committed:
Add xms-mutability

osmuller committed:
Make next link null for list

osmuller committed:
add NextLink allthough it isnt used

osmuller committed:
Add new preview api version and revert changes on existing one

osmuller committed:
Create preview version from latest stable

osmuller committed:
delete backup version

osmuller committed:
Fix json

osmuller committed:
Prettyfy

osmuller committed:
fix readme

osmuller committed:
Fix json

osmuller committed:
update yaml (unsure of fix)

osmuller committed:
Run prettifier

osmuller committed:
Add get endpoint

osmuller committed:
Change api-versions in examples

osmuller committed:
Fix swagger

osmuller committed:
Fix example

osmuller committed:
Add regex for association proxy name

osmuller committed:
prettifier

osmuller committed:
Merge branch 'main' into akv-nsp

osmuller committed:
Merge branch 'main' into akv-nsp

osmuller committed:
Change names of example

osmuller committed:
Merge branch 'akv-nsp' of https://github.com/osmuller/azure-rest-api-specs into akv-nsp

osmuller committed:
Add description to secured by perimeter

osmuller committed:
Merge branch 'main' into akv-nsp

madhura-np commented on specification/keyvault/resource-manager/Microsoft.KeyVault/preview/2024-04-01-preview/managedHsm.json:
```diff
@@ -0,0 +1,1737 @@
+    "MHSMGeoReplicatedRegion": {
+      "type": "object",
+      "properties": {
+        "name": {
```
>name [](http://example.com/codeflow?start=9&length=4)

Can there be a format applied here?

osmuller replied:
Same as other comment, this is MHSM code that we dont own and I am just copying over their existing code.

madhura-np commented on specification/keyvault/resource-manager/Microsoft.KeyVault/preview/2024-04-01-preview/managedHsm.json:
```diff
@@ -0,0 +1,1737 @@
+            "$ref": "#/parameters/ManagedHsmResourceGroupName"
+          },
+          {
+            "name": "name",
```
>name [](http://example.com/codeflow?start=13&length=4)

Can a pattern be applied?

osmuller replied:
Not sure, the MHSM side is owned by another team that piggy backs on us so I wont touch their code.

madhura-np commented on specification/keyvault/resource-manager/Microsoft.KeyVault/preview/2024-04-01-preview/keyvault.json:
```diff
@@ -0,0 +1,2411 @@
+      "description": "A private link resource",
+      "type": "object"
+    },
+    "PrivateLinkResourceProperties": {
```
>PrivateLinkResourceProperties [](http://example.com/codeflow?start=5&length=29)

@TimLovellSmith  These would be good candidates to merge the commontypes into prior to merge.

TimLovellSmith replied:
They are already defined here.

[https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/privatelinks.json](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/privatelinks.json)

And so this swagger should really reference common-types one to avoid redefining the same data types.

TimLovellSmith commented on specification/keyvault/resource-manager/Microsoft.KeyVault/preview/2024-04-01-preview/keyvault.json:
```diff
@@ -0,0 +1,2411 @@
+      "description": "A of network security perimeter associations",
+      "type": "object"
+    },
+    "NetworkSecurityPerimeterAssociationsConfigurationResult": {
```
>NetworkSecurityPerimeterAssociationsConfigurationResult [](http://example.com/codeflow?start=5&length=55)

For NSP, the hope is that, we can refactor to $ref of these definitions soon

[https://github.com/Azure/azure-rest-api-specs/pull/28958](https://github.com/Azure/azure-rest-api-specs/pull/28958)

TimLovellSmith commented on specification/keyvault/resource-manager/Microsoft.KeyVault/preview/2024-04-01-preview/keyvault.json:
```diff
@@ -0,0 +1,2411 @@
+      "description": "The name of the key vault.",
+      "x-ms-parameter-location": "method"
+    },
+    "PrivateEndpointConnectionName": {
```
>PrivateEndpointConnectionName [](http://example.com/codeflow?start=5&length=29)

can be found and reused from common-types here

[https://github.com/Azure/azure-rest-api-specs/blob/a542d7d9867b3456a99b076b8040e35fd80c0425/specification/common-types/resource-management/v5/privatelinks.json#L181C6-L181C35](https://github.com/Azure/azure-rest-api-specs/blob/a542d7d9867b3456a99b076b8040e35fd80c0425/specification/common-types/resource-management/v5/privatelinks.json#L181C6-L181C35)

TimLovellSmith commented on specification/keyvault/resource-manager/Microsoft.KeyVault/preview/2024-04-01-preview/keyvault.json:
```diff
@@ -0,0 +1,2411 @@
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/AssociationProxyNameParameter"
```
>AssociationProxyNameParameter [](http://example.com/codeflow?start=34&length=29)

This looks cognate with the (coming soon) NetworkSecurityPerimeterConfigurationNameParameter

mentat9 commented:
Looks like latest iteration has already been reviewed by @madhura-np and @TimLovellSmith, but the ARMChangesRequested label was never set. Doing it now.

TimLovellSmith commented:
@osmuller Please refactor by using references to the [new] common types NSP definitions, letting me know of any blockers

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/networksecurityperimeter.json

osmuller committed:
Merge branch 'main' into akv-nsp

osmuller committed:
Use common file

osmuller committed:
Address feedback

osmuller committed:
Merge branch 'main' into akv-nsp

osmuller commented:
@mentat9 @madhura-np Not sure if you guys are the ones that can now approve/review. Tim told me he wont be able to review this PR for a while so I was wondering if you could help me out here. Thanks!

mentat9 commented:
> @mentat9 @madhura-np Not sure if you guys are the ones that can now approve/review. Tim told me he wont be able to review this PR for a while so I was wondering if you could help me out here. Thanks!

@osmuller - Yes, I'm on-call this week and will be helping you.

Couple of open comments remain, please address/respond:
https://github.com/Azure/azure-rest-api-specs/pull/28893/files#r1597289296
https://github.com/Azure/azure-rest-api-specs/pull/28893/files#r1597289081

osmuller commented:
@mentat9 thanks Chris, just addressed the comments. Basically that is a side of the swagger that is owned by another team so I wont touch their code this iteration. This is just to add the NSP stuff so that a Policy can be created for SFI Wave2

osmuller committed:
Merge branch 'main' into akv-nsp

osmuller committed:
Merge branch 'main' into akv-nsp

osmuller commented:
/pr RequestMerge

TimLovellSmith commented:
🎉
